docs/rl-algorithms/ppo.md

```diff
@@ -712,6 +716,57 @@ Tracked experiments and game play videos:
+poetry run pip install --upgrade "jax[cuda]==0.3.17" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
+python cleanrl/ppo_atari_envpool_xla_jax_scan.py --help
+python cleanrl/ppo_atari_envpool_xla_jax_scan.py --env-id Breakout-v5
+```
+
+### Explanation of the logged metrics
+
+See [related docs](/rl-algorithms/ppo/#explanation-of-the-logged-metrics) for `ppo.py`. Just like in [ppo_atari_envpool_xla_jax.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_atari_envpool_xla_jax.py), we omit logging `losses/old_approx_kl` and `losses/clipfrac` for brevity. The metrics are the same as those in [ppo_atari_envpool_xla_jax.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_atari_envpool_xla_jax.py).
+
+### Implementation details
+
+[ppo_atari_envpool_xla_jax_scan.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_atari_envpool_xla_jax_scan.py) is a clone of [ppo_atari_envpool_xla_jax.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_atari_envpool_xla_jax.py) that replaces the python loops with native `jax.scan`.
+
+### Experiment results
+
+To run benchmark experiments, see :material-github: [benchmark/ppo.sh](https://github.com/vwxyzjn/cleanrl/blob/master/benchmark/ppo.sh). Specifically, execute the following command:
+
+<script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Fvwxyzjn%2Fcleanrl%2Fblob%2Fmaster%2Fbenchmark%2Fppo.sh%23L108-L113&style=github&type=code&showBorder=on&showLineNumbers=on&showFileMeta=on&showFullPath=on&showCopy=on"></script>
+
+Below are the average episodic returns for `ppo_atari_envpool_xla_jax_scan.py` in 3 atari games. It has the same sample efficiency as `ppo_atari_envpool_xla_jax.py`.
+
+|              | ppo_atari_envpool_xla_jax_scan ({'tag': ['pr-328']})   | baselines-ppo2-cnn ({})   | ppo_atari_envpool_xla_jax_truncation ({})   |
+|:-------------|:-------------------------------------------------------|:--------------------------|:--------------------------------------------|
+| BeamRider-v5 | 2899.62 ± 482.12                                       | 2835.71 ± 387.92          | 3133.78 ± 293.02                            |
+| Breakout-v5  | 451.27 ± 45.52                                         | 405.73 ± 11.47            | 465.90 ± 14.30                              |
+| Pong-v5      | 20.37 ± 0.20                                           | 20.45 ± 0.81              | 20.62 ± 0.18                                |
+
+???+ info
+
+    The speed of this variant and [ppo_atari_envpool_xla_jax.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_atari_envpool_xla_jax.py) are very similar but the compilation time is reduced significantly (see [vwxyzjn/cleanrl#328](https://github.com/vwxyzjn/cleanrl/pull/328#issuecomment-1340474894)). In the following learning curve, the speed increase comes from the fact that better hardware were used.
+
+Learning curves:
+
+![](../ppo/ppo_atari_envpool_xla_jax_scan/compare.png)
```

Comment:
Yeah, the hardware is a bit of an issue. Would you mind doing two additional things?

1 ) run the `ppo_atari_envpool_xla_jax.py` at your machine to align the hardware settings:

```
export WANDB_ENTITY=openrlbenchmark
python -m cleanrl_utils.benchmark \
    --env-ids Pong-v5 BeamRider-v5 Breakout-v5 \
    --command "poetry run python cleanrl/ppo_atari_envpool_xla_jax.py --track --capture-video" \
    --num-seeds 3 \
    --workers 1
```

2. regenerate the learning curves using `openrlbenchmark==0.1.1a3`, which also generates the curves with time as the x-axis. Then you can compare the experiments via 
```
pip install --upgrade openrlbenchmark
python -m openrlbenchmark.rlops \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_atari_envpool_xla_jax_scan?tag=pr-328&user=51616' 'ppo_atari_envpool_xla_jax_scan?tag=pr-328&user=51616' \
    --filters '?we=openrlbenchmark&wpn=baselines&ceik=env&cen=exp_name&metric=charts/episodic_return' 'baselines-ppo2-cnn' \
    --filters '?we=openrlbenchmark&wpn=envpool-atari&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_atari_envpool_xla_jax_truncation?user=costa-huang' \
    --env-ids BeamRider-v5 Breakout-v5  Pong-v5 \
    --check-empty-runs False \
    --ncols 3 \
    --ncols-legend 2 \
    --output-filename compare \
    --scan-history \
    --report
```

and maybe in the docs we can specify the hardware differences used by `costa-huang` and `51616` :)

Reply:
For some reason, the runs with the python loops didn't get the `pr-328` tag. So this filter doesn't work
`--filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_atari_envpool_xla_jax_scan?tag=pr-328&user=51616' 'ppo_atari_envpool_xla_jax?tag=pr-328&user=51616'`

It gives the following error:
```
{
│   'wandb_project_name': 'cleanrl',
│   'wandb_entity': 'openrlbenchmark',
│   'custom_env_id_key': 'env_id',
│   'custom_exp_name': 'exp_name',
│   'metric': 'charts/avg_episodic_return'
}
========= ppo_atari_envpool_xla_jax_scan?tag=pr-328&user=51616
========= ppo_atari_envpool_xla_jax?tag=pr-328&user=51616
{
│   'wandb_project_name': 'baselines',
│   'wandb_entity': 'openrlbenchmark',
│   'custom_env_id_key': 'env',
│   'custom_exp_name': 'exp_name',
│   'metric': 'charts/episodic_return'
}
========= baselines-ppo2-cnn
{
│   'wandb_project_name': 'envpool-atari',
│   'wandb_entity': 'openrlbenchmark',
│   'custom_env_id_key': 'env_id',
│   'custom_exp_name': 'exp_name',
│   'metric': 'charts/avg_episodic_return'
}
========= ppo_atari_envpool_xla_jax_truncation?user=costa-huang
Traceback (most recent call last):
  File "/home/tan/miniconda3/envs/cleanrl/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/tan/miniconda3/envs/cleanrl/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/tan/miniconda3/envs/cleanrl/lib/python3.7/site-packages/openrlbenchmark/rlops.py", line 356, in <module>
    time_unit=args.time_unit,
  File "/home/tan/miniconda3/envs/cleanrl/lib/python3.7/site-packages/openrlbenchmark/rlops.py", line 161, in compare
    {(runsets[idx].report_runset.name, runsets[idx].runs[0].config[runsets[idx].exp_name]): runsets[idx].color}
  File "/home/tan/miniconda3/envs/cleanrl/lib/python3.7/site-packages/wandb/apis/public.py", line 1053, in __getitem__
    return self.objects[index]
IndexError: list index out of range
```

Not sure if this is an intended behavior. Anyway, I manually added the tag and the command ran just fine.

Reply:
Hmm that was weird... The `cleanrl_utils.benchmark` utility should have autotagged.

Reply:
docs updated

Reply:
<img width="1418" alt="image" src="https://user-images.githubusercontent.com/5555347/208962803-9e0276e7-3a90-47cd-bcd5-c14cb6477cb3.png">

I actually don't quite get it... How come your experiments are like 30% faster than mine? Is it because 2080 TI is faster than 3060 TI? What is your CPU? That could also be a factor.

Reply:
My cpu is Ryzen 5950x. Maybe that’s factor too.